### PR TITLE
Use UInt32Math in JavascriptString::ToCaseCore

### PR DESCRIPTION
**Pull request by jackhorton**

### CODE REVIEW

File: lib/Runtime/Library/JavascriptString.cpp [hunk 1]
```diff
@@ -2303,7 +2303,7 @@ namespace Js
         }
         else
         {
-            charcount_t bufferLength = requiredStringLength + 1;
```
[the remark this answers is not in the document]
Comment (jackhorton) [reply lib/Runtime/Library/JavascriptString.cpp]: We were investigating this last night, and I'll update the PR to account for that. In advance, do you know what the status quo is for making tests that check OOM behavior? They're somewhat slow and can probably be flaky

**Comment by jackhorton** [comment]
ping for reviewers, I have added a lot more here to ensure safety.

File: lib/Runtime/Library/JavascriptString.cpp [hunk 2]
```diff
@@ -2282,33 +2282,69 @@ namespace Js
...
+        {
+            if (requiredStringLength == 1)
+            {
+                // don't create a new string in case we may have cached this string earlier
+                return scriptContext->GetLibrary()->GetCharStringCache().GetStringForChar(guessBuffer[0]);
+            }
+            else
+            {
+                // use requiredStringLength instead of guessBufferLength because the string can get shorter
+                return JavascriptString::NewWithBuffer(guessBuffer, requiredStringLength, scriptContext);
```
Comment (MSLaguana) [inline lib/Runtime/Library/JavascriptString.cpp]: Does this mean that if there is `NoError`, we know for sure that `requiredStringLength <= guessBufferLength`?

Comment (jackhorton) [reply lib/Runtime/Library/JavascriptString.cpp]: Well... In theory, yes. Probably good to check it anyways!

**Comment by jackhorton** [comment]
Yeah, its not the best. As i mentioned in the comment I did add, ChangeStringLinguisticCase should really be recycler-aware such that it can have the signature 

```cpp
void ChangeStringLinguisticCase(_In_ Allocator *alloc, _In_ char16 *src, _In_ charcount_t srcLength, _Out_ char16 **dest, _Out_ charcount_t *length);
```

And not have to force the callers to worry about platform differences. That is sortof the whole point of the PlatformAgnostic API, but until it knows how to allocate in a known-good memory pool and throw out of memory and such, I would prefer to leave the API as is. I hope this will change when we take WinGlob out so that I can move PlatformAgnostic back to lib\Common so that it knows about Chakra.Memory and can be supplied global error functions like how Utf8Codex works.

**Pull request merged**